server/app/services/rpc/node_service_pod_handler.rb

```diff
@@ -66,6 +63,27 @@ def set_state(id, pod)
       {}
     end
 
+    # @param id [String]
+    # @param pod [Hash]
+    # @param hook [Hash]
+    def mark_oneshot_hook(id, pod, hook)
+      node = @grid.host_nodes.find_by(node_id: id)
+      raise 'Node not found' unless node
+
+      service_instance = node.grid_service_instances.find_by(
+        grid_service_id: pod['service_id'], instance_number: pod['instance_number']
+      )
+      raise 'Instance not found' unless service_instance
+
+      oneshot_hook = service_instance.grid_service.hooks.to_a.find{ |h|
+        h.oneshot && h.type == hook['type'] && h.cmd == hook['cmd']
```

Comment:
Not sure if we can trust that embedded `hooks` collection does not get wiped and replaced with new id?

Reply:
This should have been fixed in  #2097, and there's a [spec](https://github.com/kontena/kontena/blob/6b134c6f96d44ff05fe0056f188d8979f20a62f6/server/spec/mutations/grid_services/update_spec.rb#L368-L386) for it.

Besides, the `hook['id']` would only be used briefly between the `list` -> `mark_oneshot_hook` RPCs.

And the same `GridServiceHook` object also contains the `done` array, if the embedded object gets wiped, then we'll have bigger problems with the `oneshot` hooks anyways.

Reply:
This really needs the `hook['id']`, there can be multiple identical commands in the service hooks, and this kind of `hooks.find{...}` will only mark the first such command:

```yaml
stack: test/hooks-oneshot
services:
  redis:
    image: redis
    hooks:
      post_start:
        - name: test-first
          cmd: date && echo "first"
          instances: '*'
          oneshot: false
        - name: test-oneshot1
          cmd: date && echo "pre-sleep"
          instances: '*'
          oneshot: true
        - name: test-sleep1
          cmd: sleep 5
          instances: '*'
          oneshot: true
        - name: test-echo2
          cmd: date && echo "mid-sleep"
          instances: '*'
          oneshot: true
        - name: test-sleep2
          cmd: sleep 5
          instances: '*'
          oneshot: true
        - name: test-echo3
          cmd: date && echo "post-sleep"
          instances: '*'
          oneshot: true
        - name: test-last
          cmd: date && echo "last"
          instances: '*'
          oneshot: false
```

Initial deploy:

```
D, [2017-09-20T12:02:16.747939 #1] DEBUG -- Kontena::ServicePods::Creator: creating container: hooks-oneshot.redis-1
D, [2017-09-20T12:02:17.024074 #1] DEBUG -- Kontena::ServicePods::Creator: container created: hooks-oneshot.redis-1
I, [2017-09-20T12:02:18.504467 #1]  INFO -- Kontena::ServicePods::Creator: service started: hooks-oneshot/redis-1
I, [2017-09-20T12:02:18.661459 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "first"
I, [2017-09-20T12:02:19.038625 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "pre-sleep"
I, [2017-09-20T12:02:19.298197 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: sleep 5
I, [2017-09-20T12:02:24.727551 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "mid-sleep"
I, [2017-09-20T12:02:24.842957 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: sleep 5
I, [2017-09-20T12:02:29.964046 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "post-sleep"
I, [2017-09-20T12:02:30.086937 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "last"
```

Restart the agent to clear out the `LifecycleHookManager@oneshot_cache`, and re-deploy the service:

```
I, [2017-09-20T12:03:02.891027 #1]  INFO -- Kontena::ServicePods::Creator: removing previous version of service: hooks-oneshot/redis-1
D, [2017-09-20T12:03:03.406122 #1] DEBUG -- Kontena::ServicePods::Creator: creating container: hooks-oneshot.redis-1
D, [2017-09-20T12:03:03.496747 #1] DEBUG -- Kontena::ServicePods::Creator: container created: hooks-oneshot.redis-1
I, [2017-09-20T12:03:03.962722 #1]  INFO -- Kontena::ServicePods::Creator: service started: hooks-oneshot/redis-1
I, [2017-09-20T12:03:04.051490 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "first"
I, [2017-09-20T12:03:04.276366 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: sleep 5
I, [2017-09-20T12:03:09.404685 #1]  INFO -- Kontena::ServicePods::LifecycleHookManager: running post_start hook: date && echo "last"
```

The second `sleep 5` hook gets executed again, even though it's `oneshot`

Reply:
Turns out this wasn't actually because of the way that the `find` works here, it was because the agent has a similar problem, and doesn't even mark the second `sleep 5` hook: https://github.com/kontena/kontena/pull/2701/files#r139954994